File: app/src/main/java/com/example/mike0/redditkotlinactivity/data/model/Child.java

package com.example.mike0.redditkotlinactivity.data.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

public class Child implements Parcelable
{

    @SerializedName("kind")
    @Expose
    private String kind;
    @SerializedName("data")
    @Expose
    private Post data;
    public final static Parcelable.Creator<Child> CREATOR = new Creator<Child>() {


        @SuppressWarnings({
            "unchecked"
        })
        public Child createFromParcel(Parcel in) {
            return new Child(in);
        }

        public Child[] newArray(int size) {
            return (new Child[size]);
        }

    }
    ;

    protected Child(Parcel in) {
        this.kind = ((String) in.readValue((String.class.getClassLoader())));
        this.data = ((Post) in.readValue((Post.class.getClassLoader())));
    }

    public Child() {
    }

    public String getKind() {
        return kind;
    }

    public void setKind(String kind) {
        this.kind = kind;
    }

    public Post getData() {
        return data;
    }

    public void setData(Post data) {
        this.data = data;
    }

    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(kind);
        dest.writeValue(data);
    }

    public int describeContents() {
        return  0;
    }

}

File: app/src/main/java/com/example/mike0/redditkotlinactivity/data/model/Child.kt
package com.example.mike0.redditkotlinactivity.data.model

import android.os.Parcel
import android.os.Parcelable

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName

class Child : Parcelable {

    @SerializedName("kind")
    @Expose
    var kind: String? = null
    @SerializedName("data")
    @Expose
    var data: Post? = null

    protected constructor(`in`: Parcel) {
        this.kind = `in`.readValue(String::class.java.classLoader) as String
        this.data = `in`.readValue(Post::class.java.classLoader) as Post
    }

    companion object {
        val CREATOR: Parcelable.Creator<Child> = object : Parcelable.Creator<Child> {

            override fun createFromParcel(`in`: Parcel): Child {
                return Child(`in`)
            }

            override fun newArray(size: Int): Array<Child?> {
                return arrayOfNulls(size)
            }
        }
    }

    constructor() {}

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(kind)
        dest.writeValue(data)
    }

    override fun describeContents(): Int {
        return 0
    }
}